### PR TITLE
metrics: optimize launch delay computation and reduce potential repeatedly metering.

### PR DESCRIPTION
**Pull request by SimonCqk**
1. reduce potential repeatedly metering of JobFailure and JobRestarting.
2. optimize launch delay computation.

### CODE REVIEW

File: pkg/job_controller/job.go [hunk 1]
```diff
@@ -251,6 +251,11 @@ func (jc *JobController) ReconcileJobs(
 		return result, err
 	}
 
+	// Metring first pod launch delay when job state transit from created to running.
+	if !commonutil.IsRunning(*oldStatus) && commonutil.IsRunning(jobStatus) {
```
Comment (jian-he) [inline pkg/job_controller/job.go]: The comment says from "created" , but the code says !isRunning, 
the job could be transitioned from Restarting too

File: pkg/metrics/job_metrics.go [hunk 2]
```diff
@@ -136,15 +136,28 @@ func (m *JobMetrics) RestartInc() {
...
+		}
+		if earliestTime == nil || readyCond.LastTransitionTime.Before(earliestTime) {
+			earliestTime = &readyCond.LastTransitionTime
+		}
+	}
+	if earliestTime == nil {
 		return
 	}
-	delay := metav1.Now().Time.Sub(status.StartTime.Time).Seconds()
+	delay := earliestTime.Time.Sub(status.StartTime.Time).Seconds()
```
Comment (jian-he) [inline pkg/metrics/job_metrics.go]: The StartTime for xgboost and xdl job currently are defined as ALL workers running. 
In that case, the earliestTime before the StartTime, may be we can use the job object's creationTimestamp

File: pkg/metrics/job_metrics.go [hunk 3]
```diff
@@ -154,16 +167,21 @@ func (m *JobMetrics) FirstPodLaunchDelaySeconds(job metav1.Object, status v1.Job
 }
 
 func (m *JobMetrics) AllPodsLaunchDelaySeconds(pods []*corev1.Pod, job metav1.Object, status v1.JobStatus) {
-	if status.StartTime == nil {
+	if !util.IsRunning(status) || status.StartTime == nil {
```
Comment (jian-he) [inline pkg/metrics/job_metrics.go]: Similarly, the StartTime definition is different for xdl,xgboost jobs.  we may use job's creationTimestamp

File: pkg/job_controller/job.go [hunk 4]
```diff
@@ -251,6 +251,11 @@ func (jc *JobController) ReconcileJobs(
...
+		jc.Metrics.FirstPodLaunchDelaySeconds(activePods, metaObject, jobStatus)
+	}
+
 	jc.Metrics.AllPodsLaunchDelaySeconds(pods, metaObject, jobStatus)
```
Comment (jian-he) [inline pkg/job_controller/job.go]: looks like this AllPodsLaunchDelaySeconds will be called again and again,  we only need it for the first All pods start running

File: pkg/job_controller/job.go (hunk 1 above)
Comment (SimonCqk) [reply pkg/job_controller/job.go]: done, the old Status must be `Created`.

File: pkg/metrics/job_metrics.go (hunk 2 above)
Comment (SimonCqk) [reply pkg/metrics/job_metrics.go]: done, creationTimestamp is a more appropriate choice.

File: pkg/job_controller/job.go (hunk 4 above)
Comment (SimonCqk) [reply pkg/job_controller/job.go]: done, check before-after number of active pods.

File: pkg/job_controller/job.go [hunk 5]
```diff
@@ -251,7 +251,17 @@ func (jc *JobController) ReconcileJobs(
...
 
-	jc.Metrics.AllPodsLaunchDelaySeconds(pods, metaObject, jobStatus)
+	// Metering first pod launch delay when job state transit from created to running.
+	if commonutil.IsCreated(*oldStatus) && commonutil.IsRunning(jobStatus) {
+		jc.Metrics.FirstPodLaunchDelaySeconds(activePods, metaObject, jobStatus)
+	}
+
+	// Metring all pods launch delay when latest pods are all active after reconciled, and previous
+	// job status has not reached a all-active state.
+	if (k8sutil.GetTotalAvtiveReplicas(jobStatus.ReplicaStatuses) == totalReplicas) &&
```
Comment (jian-he) [inline pkg/job_controller/job.go]: If the there are pods failed and re-created while the job is running.
these two conditions can also be true I think .

Comment (SimonCqk) [reply pkg/job_controller/job.go]: done, with extra  condition `!IsRestarting(oldStatus)`.

**Pull request merged**